compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs

```diff
@@ -198,32 +232,59 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         }
     }
 
+    /// Try our best to approximate when adding `dyn` would be helpful for a bare
+    /// trait object.
+    ///
+    /// Right now, this is if the type is either directly nested in another ty,
+    /// or if it's in the tail field within a struct. This approximates what the
+    /// user would've gotten on edition 2015, except for the case where we have
+    /// an *obvious* knock-on `Sized` error.
     fn maybe_suggest_dyn_trait(
         &self,
         self_ty: &hir::Ty<'_>,
-        label: &str,
         sugg: Vec<(Span, String)>,
         diag: &mut Diag<'_>,
     ) -> bool {
         let tcx = self.tcx();
-        let parent_id = tcx.hir_get_parent_item(self_ty.hir_id).def_id;
-        let parent_item = tcx.hir_node_by_def_id(parent_id).expect_item();
 
-        // If the parent item is an enum, don't suggest the dyn trait.
-        if let hir::ItemKind::Enum(..) = parent_item.kind {
-            return false;
-        }
+        // Look at the direct HIR parent, since we care about the relationship between
```

Comment:
This was also `expect_item`ing unnecessarily. I took the logic and made it focus on the relationship between the ty and its parent *node*, so we don't suggest turning:

```
struct Foo { f: Box<Any>, /* more fields */ }
```

Into this:

```
struct Foo<T: Any> { f: Box<T>, /* more fields */ }
```

My general logic here is that if the ty's hir node parent is another ty, then it's probably just a misspelled `dyn Trait`. There is no perfect heuristic here, but I think it's *less* invasive to suggest the most naive thing here, which is just to insert `dyn`, unless we're very confident it will be wrong.